aten/src/ATen/ThreadLocalDebugInfo.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <c10/macros/Export.h>
+
+#include <memory>
+#include <string>
+
+namespace at {
+
+class CAFFE2_API ThreadLocalDebugInfoBase {
+ public:
+  ThreadLocalDebugInfoBase() {}
+  virtual ~ThreadLocalDebugInfoBase() {}
+};
```

Comment:
Non-blocking for this PR, but something to think about: the single thread local field that you've installed here is not compositional: if one client has DebugInfoA and another has DebugInfoB, there is no way for both of them to actually propagate both of them. Because of this lack of compositionality, I don't really see the point of abstracting the class in this way: it would probably be better to just have a single monolithic class that you put all of the stuff you want in, and if it becomes unmanageable at some later point in time, *then* add whatever abstraction you need to handle the debug information in a more scalable way. Basically, this is not one of those cases where generalizing early saves you pain later, so let's generalize as late as possible!

Reply:
Unless, I suppose, what's actually happening is you're extending this class from inside fbcode, because the debug info you want to store doesn't exist inside PyTorch proper. But this isn't mentioned anywhere inside the PR; I'm just guessing ;)

Reply:
I guess this is implied by the location of the documentation you added in docs/source/notes/large_scale_deployments.rst

Reply:
I guess what I would like to see, is something in big, blinky, bold red letters, saying **DO NOT USE THIS IF YOU ARE A LIBRARY AUTHOR, THIS IS NOT FOR YOU**

Reply:
@ezyang - yes, large_scale_deployments.rst explains the main use case and the test I added demoes it - we want to transparently pass something that does not exist within pure PT from the user to the user-authored hooks (e.g. logging). Not sure about big, blinky, bold red letters, but I'll add more into the comment in the code :)